src/preamble.js

```diff
@@ -164,27 +164,37 @@ function ccall (ident, returnType, argTypes, args, opts) {
     }
   }
   var ret = func.apply(null, cArgs);
-#if ASSERTIONS
 #if EMTERPRETIFY_ASYNC
+  if (opts && opts.callback) opts.async = true;
+#if ASSERTIONS
   if ((!opts || !opts.async) && typeof EmterpreterAsync === 'object') {
     assert(!EmterpreterAsync.state, 'cannot start async op with normal JS calling ccall');
   }
-  if (opts && opts.async) assert(!returnType, 'async ccalls cannot return values');
-#endif
+  if (opts && opts.async && returnType) assert(opts.callback, 'async ccalls need a callback to return values');
+#endif
+  if (opts && opts.async && EmterpreterAsync.state) {
```

Comment:
why does this check `EmterpreterAsync.state`?

Reply:
This worries me. Can we instead assert on an async operation happening when a call is declared as async?

Reply:
You can't always know what it will do. For example, in my project I use this to execute a VM. It could pause and wait for input, or it could just output some text and finish.

Reply:
Hmm, what purpose does `opts.async` actually serve? Previously it was used to restore the stack (see below, lines 179-186, but that code is replaced by this PR. Arguably it should always have been checking `EmterpreterAsync.state` as that's more robust. With this PR `opts.async` is now only being used in assertions, so should it be removed? This would mean moving the 'async ccalls need a callback to return values' assertion inside of this if block.

Reply:
Right, but what kind of strange behaviour are you thinking of? With this PR, nothing would break if you left out the opt (if you're not expecting a return value), the only thing that might be different is the timing of when the Emterpreter actually finishes. It's the reverse of our discussion below about delaying the callback call ;) If I'm right that nothing would break but you both still want to keep the opt, could it be changed to console.warn?

Reply:
Ok, I see now what you mean. So the async opt wasn't actually necessary even before this PR. Adding the callback and removing the async opt are really independent changes, but I can add it to the PR.

Reply:
The idea was that specifying a callback means explicitly saying that the function is async. In case there is no return value, you could say that it doesn't matter if the call is async or not.
But now that I think about it, that's a bad idea, because one might except that an action is completed synchronously when calling without setting async.

Reply:
Right, `opts.async` is good for indicating user intention, and for assertions, but it doesn't actually serve anything in the code. As it is now, we make everyone provide the parameter even when it's not really necessary.

My compromise recommendation would be to make it non-required, but change the "cannot start async op with normal JS calling ccall" assertion to a warning if you don't provide it. And still only in ASSERTIONS mode so those who know what they're doing can leave it out if they prefer.

Reply:
But is it really that much of an issue if you have to add the `async` opt to async ccalls? Adding the option means saying "I know what I'm doing".

Reply:
You're right, it's not much of an issue. But when it doesn't really serve a purpose, why are we having to prove to our computer that we know what we're doing? Ideologically I'd just prefer to not have to provide parameters which don't do anything. :)

Reply:
I agree with

> But now that I think about it, that's a bad idea, because one might except that an action is completed synchronously when calling without setting async.

It maybe is not necessary to be explicit in that case, but it's safer. So I like `async: function ..` when there is a return value, and `async: true` (or any other non-falsey value) for when there isn't.